File: python/src/pi_mono/agent/harness/messages.py
"""Message formatting utilities for agent harness."""

from typing import Any
import time
from datetime import datetime

from pi_mono.agent import AgentMessage
from pi_mono.ai.types import ImageContent, TextContent

COMPACTION_SUMMARY_PREFIX = """The conversation history before this point was compacted into the following summary:

<summary>
"""

COMPACTION_SUMMARY_SUFFIX = """
</summary>
"""

BRANCH_SUMMARY_PREFIX = """The following is a summary of a branch that this conversation came back from:

<summary>
"""

BRANCH_SUMMARY_SUFFIX = """</summary>"""


def parse_iso_timestamp(timestamp_str: str) -> int:
    try:
        t = timestamp_str.replace("Z", "+00:00")
        dt = datetime.fromisoformat(t)
        return int(dt.timestamp() * 1000)
    except Exception:
        return int(time.time() * 1000)


def create_user_message(text: str, images: list[ImageContent] | None = None) -> AgentMessage:
    """Create a user message with optional images."""
    content: list[TextContent | ImageContent] = [{"type": "text", "text": text}]
    if images:
        content.extend(images)
    return {
        "role": "user",
        "content": content,
        "timestamp": int(time.time() * 1000),
    }


class BashExecutionMessage:
    role: str = "bashExecution"
    command: str
    output: str
    exit_code: int | None = None
    cancelled: bool
    truncated: bool
    full_output_path: str | None = None
    timestamp: int
    exclude_from_context: bool = False

    def __init__(
        self,
        command: str,
        output: str,
        exit_code: int | None,
        cancelled: bool,
        truncated: bool,
        full_output_path: str | None = None,
        timestamp: int | None = None,
        exclude_from_context: bool = False,
    ):
        self.command = command
        self.output = output
        self.exit_code = exit_code
        self.cancelled = cancelled
        self.truncated = truncated
        self.full_output_path = full_output_path
        self.timestamp = timestamp or 0
        self.exclude_from_context = exclude_from_context


class CustomMessage:
    role: str = "custom"
    custom_type: str
    content: str | list[TextContent | ImageContent]
    display: bool
    details: Any | None = None
    timestamp: int

    def __init__(
        self,
        custom_type: str,
        content: str | list[TextContent | ImageContent],
        display: bool,
        details: Any | None = None,
        timestamp: int | None = None,
    ):
        self.custom_type = custom_type
        self.content = content
        self.display = display
        self.details = details
        self.timestamp = timestamp or 0


class BranchSummaryMessage:
    role: str = "branchSummary"
    summary: str
    from_id: str
    timestamp: int

    def __init__(self, summary: str, from_id: str, timestamp: str):
        self.summary = summary
        self.from_id = from_id
        self.timestamp = parse_iso_timestamp(timestamp)


class CompactionSummaryMessage:
    role: str = "compactionSummary"
    summary: str
    tokens_before: int
    timestamp: int

    def __init__(self, summary: str, tokens_before: int, timestamp: str):
        self.summary = summary
        self.tokens_before = tokens_before
        self.timestamp = parse_iso_timestamp(timestamp)


def bash_execution_to_text(msg: Any) -> str:
    """Convert bash execution message to text representation."""
    command = getattr(msg, "command", msg.get("command") if isinstance(msg, dict) else "")
    output = getattr(msg, "output", msg.get("output") if isinstance(msg, dict) else "")
    cancelled = getattr(msg, "cancelled", msg.get("cancelled") if isinstance(msg, dict) else False)
    exit_code = getattr(
        msg,
        "exit_code",
        (
            getattr(msg, "exitCode", msg.get("exitCode", msg.get("exit_code")))
            if isinstance(msg, dict)
            else None
        ),
    )
    truncated = getattr(msg, "truncated", msg.get("truncated") if isinstance(msg, dict) else False)
    full_output_path = getattr(
        msg,
        "full_output_path",
        (
            getattr(msg, "fullOutputPath", msg.get("fullOutputPath", msg.get("full_output_path")))
            if isinstance(msg, dict)
            else None
        ),
    )

    text = f"Ran `{command}`\n"
    if output:
        text += f"```\n{output}\n```"
    else:
        text += "(no output)"
    if cancelled:
        text += "\n\n(command cancelled)"
    elif exit_code is not None and exit_code != 0:
        text += f"\n\nCommand exited with code {exit_code}"
    if truncated and full_output_path:
        text += f"\n\n[Output truncated. Full output: {full_output_path}]"
    return text


def create_branch_summary_message(summary: str, from_id: str, timestamp: str) -> dict:
    """Create a branch summary message."""
    return {
        "role": "branchSummary",
        "summary": summary,
        "fromId": from_id,
        "timestamp": parse_iso_timestamp(timestamp),
    }


def create_compaction_summary_message(summary: str, tokens_before: int, timestamp: str) -> dict:
    """Create a compaction summary message."""
    return {
        "role": "compactionSummary",
        "summary": summary,
        "tokensBefore": tokens_before,
        "timestamp": parse_iso_timestamp(timestamp),
    }


def create_custom_message(
    custom_type: str,
    content: str | list[dict],
    display: bool,
    details: Any | None,
    timestamp: str,
) -> dict:
    """Create a custom message."""
    return {
        "role": "custom",
        "customType": custom_type,
        "content": content,
        "display": display,
        "details": details,
        "timestamp": parse_iso_timestamp(timestamp),
    }


def create_failure_message(model: Any, error: Exception, aborted: bool) -> dict:
    """Create a failure message for when an agent run fails."""
    from pi_mono.agent.harness.types import AgentHarnessError

    if isinstance(error, AgentHarnessError):
        message_text = error.message
    elif isinstance(error, Exception) and error.args and isinstance(error.args[0], str):
        message_text = error.args[0]
    else:
        message_text = str(error)

    model_dict = model if isinstance(model, dict) else {}
    model_api = getattr(model, "api", model_dict.get("api", "unknown"))
    model_provider = getattr(model, "provider", model_dict.get("provider", "unknown"))
    model_id = getattr(model, "id", model_dict.get("id", "unknown"))

    return {
        "role": "assistant",
        "content": [],
        "api": model_api,
        "provider": model_provider,
        "model": model_id,
        "stopReason": "aborted" if aborted else "error",
        "errorMessage": message_text,
        "timestamp": int(time.time() * 1000),
        "usage": {
            "input": 0,
            "output": 0,
            "cacheRead": 0,
            "cacheWrite": 0,
            "totalTokens": 0,
            "cost": {"input": 0, "output": 0, "cacheRead": 0, "cacheWrite": 0, "total": 0},
        },
    }


def convert_to_llm(messages: list[dict]) -> list[dict]:
    """Convert agent messages to LLM-compatible format."""
    result: list[dict] = []
    for m in messages:
        role = m.get("role")
        if role == "bashExecution":
            if m.get("excludeFromContext"):
                continue
            result.append(
                {
                    "role": "user",
                    "content": [{"type": "text", "text": bash_execution_to_text(m)}],
                    "timestamp": m.get("timestamp", 0),
                }
            )
        elif role == "custom":
            content = m.get("content")
            if isinstance(content, str):
                content = [{"type": "text", "text": content}]
            result.append(
                {
                    "role": "user",
                    "content": content,
                    "timestamp": m.get("timestamp", 0),
                }
            )
        elif role == "branchSummary":
            result.append(
                {
                    "role": "user",
                    "content": [
                        {
                            "type": "text",
                            "text": BRANCH_SUMMARY_PREFIX + m["summary"] + BRANCH_SUMMARY_SUFFIX,
                        }
                    ],
                    "timestamp": m.get("timestamp", 0),
                }
            )
        elif role == "compactionSummary":
            result.append(
                {
                    "role": "user",
                    "content": [
                        {
                            "type": "text",
                            "text": COMPACTION_SUMMARY_PREFIX
                            + m["summary"]
                            + COMPACTION_SUMMARY_SUFFIX,
                        }
                    ],
                    "timestamp": m.get("timestamp", 0),
                }
            )
        elif role in ("user", "assistant", "toolResult"):
            result.append(m)
    return result

File: packages/agent/src/harness/messages.ts
import type { ImageContent, Message, TextContent } from "@earendil-works/pi-ai";
import type { AgentMessage } from "../types.ts";

/**
 * 向 LLM 展示压缩摘要时使用的前缀。
 * 摘要文本插入在此前缀和 {@link COMPACTION_SUMMARY_SUFFIX} 之间。
 */
export const COMPACTION_SUMMARY_PREFIX = `The conversation history before this point was compacted into the following summary:

<summary>
`;

/**
 * 向 LLM 展示压缩摘要时使用的后缀，用于闭合压缩摘要包装。
 * 与 {@link COMPACTION_SUMMARY_PREFIX} 配对使用。
 */
export const COMPACTION_SUMMARY_SUFFIX = `
</summary>`;

/**
 * 向 LLM 展示分支摘要时使用的前缀。
 * 当对话从子代理分支返回时插入分支摘要，提供该分支中已完成操作的上下文。
 * 摘要文本插入在此前缀和 {@link BRANCH_SUMMARY_SUFFIX} 之间。
 */
export const BRANCH_SUMMARY_PREFIX = `The following is a summary of a branch that this conversation came back from:

<summary>
`;

/**
 * 向 LLM 展示分支摘要时使用的后缀，用于闭合分支摘要包装。
 * 与 {@link BRANCH_SUMMARY_PREFIX} 配对使用。
 */
export const BRANCH_SUMMARY_SUFFIX = `</summary>`;

/**
 * 表示由 harness 捕获的 bash/shell 命令执行事件。
 * 通过 {@link bashExecutionToText} 转换为对 LLM 可见的 user 角色消息。
 */
export interface BashExecutionMessage {
	role: "bashExecution";
	command: string;
	output: string;
	exitCode: number | undefined;
	cancelled: boolean;
	truncated: boolean;
	fullOutputPath?: string;
	timestamp: number;
	excludeFromContext?: boolean;
}

/**
 * 表示应用自定义消息，可以包含文本、图片或任意详细数据。
 * 当 {@link display} 为 true 时，在 LLM 上下文中显示为 user 角色消息。
 *
 * @typeParam T - 可选 {@link details} 载荷的类型。
 */
export interface CustomMessage<T = unknown> {
	role: "custom";
	customType: string;
	content: string | (TextContent | ImageContent)[];
	display: boolean;
	details?: T;
	timestamp: number;
}

/**
 * 表示对话从子代理分支返回时的摘要。
 * 转换为 LLM 可见消息时，用 {@link BRANCH_SUMMARY_PREFIX} 和 {@link BRANCH_SUMMARY_SUFFIX}
 * 包装。
 */
export interface BranchSummaryMessage {
	role: "branchSummary";
	summary: string;
	fromId: string;
	timestamp: number;
}

/**
 * 表示对话压缩摘要，当上下文窗口过大时替换较早的消息。
 * 转换为 LLM 可见消息时，用 {@link COMPACTION_SUMMARY_PREFIX} 和
 * {@link COMPACTION_SUMMARY_SUFFIX} 包装。
 */
export interface CompactionSummaryMessage {
	role: "compactionSummary";
	summary: string;
	tokensBefore: number;
	timestamp: number;
}

declare module "../types.ts" {
	interface CustomAgentMessages {
		bashExecution: BashExecutionMessage;
		custom: CustomMessage;
		branchSummary: BranchSummaryMessage;
		compactionSummary: CompactionSummaryMessage;
	}
}

/**
 * 将 {@link BashExecutionMessage} 转换为对 LLM 可见的人类可读文本表示。
 * 包括执行的命令、输出（用代码块包裹）、退出码、取消状态以及
 * 适用时的截断提示。
 */
export function bashExecutionToText(msg: BashExecutionMessage): string {
	let text = `Ran \`${msg.command}\`\n`;
	if (msg.output) {
		text += `\`\`\`\n${msg.output}\n\`\`\``;
	} else {
		text += "(no output)";
	}
	if (msg.cancelled) {
		text += "\n\n(command cancelled)";
	} else if (msg.exitCode !== null && msg.exitCode !== undefined && msg.exitCode !== 0) {
		text += `\n\nCommand exited with code ${msg.exitCode}`;
	}
	if (msg.truncated && msg.fullOutputPath) {
		text += `\n\n[Output truncated. Full output: ${msg.fullOutputPath}]`;
	}
	return text;
}

/**
 * 创建一个 {@link BranchSummaryMessage}，捕获在返回主对话前子代理分支中
 * 发生的事件。
 *
 * @param summary - 分支活动的人类可读摘要。
 * @param fromId - 此摘要来源的分支标识符。
 * @param timestamp - 表示摘要创建时间的 ISO-8601 时间戳字符串。
 * @returns 一个新的 {@link BranchSummaryMessage} 实例。
 */
export function createBranchSummaryMessage(summary: string, fromId: string, timestamp: string): BranchSummaryMessage {
	return {
		role: "branchSummary",
		summary,
		fromId,
		timestamp: new Date(timestamp).getTime(),
	};
}

/**
 * 创建一个 {@link CompactionSummaryMessage}，对话历史被压缩以保持在
 * 上下文窗口限制内后的摘要。
 *
 * @param summary - 较早对话历史的压缩摘要。
 * @param tokensBefore - 压缩发生前的近似 token 数量。
 * @param timestamp - 表示压缩发生时间的 ISO-8601 时间戳字符串。
 * @returns 一个新的 {@link CompactionSummaryMessage} 实例。
 */
export function createCompactionSummaryMessage(
	summary: string,
	tokensBefore: number,
	timestamp: string,
): CompactionSummaryMessage {
	return {
		role: "compactionSummary",
		summary,
		tokensBefore,
		timestamp: new Date(timestamp).getTime(),
	};
}

/**
 * 为应出现在 LLM 对话上下文中的应用自定义内容创建一个 {@link CustomMessage}。
 *
 * @param customType - 标识自定义消息类型的字符串标签。
 * @param content - 消息内容，可以是纯文本字符串或文本/图片块数组。
 * @param display - 此消息是否应在 LLM 上下文中可见。
 * @param details - 与消息关联的可选任意载荷。
 * @param timestamp - 表示消息创建时间的 ISO-8601 时间戳字符串。
 * @returns 一个新的 {@link CustomMessage} 实例。
 */
export function createCustomMessage(
	customType: string,
	content: string | (TextContent | ImageContent)[],
	display: boolean,
	details: unknown | undefined,
	timestamp: string,
): CustomMessage {
	return {
		role: "custom",
		customType,
		content,
		display,
		details,
		timestamp: new Date(timestamp).getTime(),
	};
}

/**
 * 将 {@link AgentMessage} 对象数组转换为标准 LLM {@link Message} 格式，
 * 用于模型提供商 API。
 *
 * 角色转换规则：
 * - **bashExecution**：转换为 `user` 角色消息，命令输出通过
 *   {@link bashExecutionToText} 渲染。如果 `excludeFromContext` 为 true，
 *   则完全跳过。
 * - **custom**：转换为 `user` 角色消息，直接使用消息的
 *   内容（字符串或结构化内容数组）。
 * - **branchSummary**：转换为 `user` 角色消息，摘要文本用
 *   {@link BRANCH_SUMMARY_PREFIX}/{@link BRANCH_SUMMARY_SUFFIX} 包装。
 * - **compactionSummary**：转换为 `user` 角色消息，摘要文本用
 *   {@link COMPACTION_SUMMARY_PREFIX}/{@link COMPACTION_SUMMARY_SUFFIX} 包装。
 * - **user / assistant / toolResult**：原样传递
 *   （已经是标准 LLM 消息格式）。
 * - **所有其他角色**：过滤掉（从结果中省略）。
 *
 * @param messages - 要转换的代理消息数组。
 * @returns 适用于模型提供商的标准 LLM {@link Message} 对象数组。
 */
export function convertToLlm(messages: AgentMessage[]): Message[] {
	return messages
		.map((m): Message | undefined => {
			switch (m.role) {
				case "bashExecution":
					if (m.excludeFromContext) {
						return undefined;
					}
					return {
						role: "user",
						content: [{ type: "text", text: bashExecutionToText(m) }],
						timestamp: m.timestamp,
					};
				case "custom": {
					const content = typeof m.content === "string" ? [{ type: "text" as const, text: m.content }] : m.content;
					return {
						role: "user",
						content,
						timestamp: m.timestamp,
					};
				}
				case "branchSummary":
					return {
						role: "user",
						content: [{ type: "text" as const, text: BRANCH_SUMMARY_PREFIX + m.summary + BRANCH_SUMMARY_SUFFIX }],
						timestamp: m.timestamp,
					};
				case "compactionSummary":
					return {
						role: "user",
						content: [
							{ type: "text" as const, text: COMPACTION_SUMMARY_PREFIX + m.summary + COMPACTION_SUMMARY_SUFFIX },
						],
						timestamp: m.timestamp,
					};
				case "user":
				case "assistant":
				case "toolResult":
					return m;
				default:
					return undefined;
			}
		})
		.filter((m): m is Message => m !== undefined);
}
